### PR TITLE
Further recluster optimizations

### PR DESCRIPTION
**Pull request by aelkiss**
- Batching during recluster
- Optimize when to recluster when HTItems move clusters

### CODE REVIEW

**Comment by aelkiss** [comment]
The May full hathifiles load succeeded with the optimizations here

File: lib/cluster_ht_item.rb [hunk 1]
```diff
@@ -46,7 +46,8 @@ def delete
     Retryable.with_transaction do
       if (cluster = cluster_with_htitem(ht_item))
         Services.logger.debug "removing old htitem #{ht_item.item_id}"
-        cluster.ht_item(ht_item.item_id).delete
+        old_item = cluster.ht_item(ht_item.item_id)
+        old_item.delete
 
         # Note that technically we only need to do this if there were multiple
```
Comment (aelkiss) [inline lib/cluster_ht_item.rb:52]: This comment is out of date now

File: lib/cluster_ht_item.rb [hunk 2]
```diff
@@ -98,17 +101,39 @@ def update_or_add_ht_items(cluster)
...
+    # - There was 0 or 1 old OCN (so it couldn't have been 'glue')
+    # - old_ocns are all in concordance rules (so this item is not the 'glue')
+    # - old_ocns are a subset of the new ocns (so it remains 'glue')
+
+    old_ocns = old_ocns.to_set
     new_ocns = new_ocns.to_set
     concordance_ocns = cluster.ocn_resolutions.collect(&:ocns).flatten.to_set
 
-    !(current_ocns.subset?(new_ocns) || current_ocns.subset?(concordance_ocns))
+    !(old_ocns.count <= 1 ||
```
Comment (aelkiss) [inline lib/cluster_ht_item.rb:119]: Worth pondering more whether we should extract the common logic here and with `move_needs_recluster?` - is there a named concept here we can extract or is the similarity incidental?

Comment (jsteverman) [reply lib/cluster_ht_item.rb:119]: I *think* this is redundant. `move_needs_cluster?` is simply `update_needs_recluster?` with `new_ocns` as an empty set. In that case `old_ocns.subset?(new_ocns)` will always be false.

Changing `update_needs_recluster?(cluster, old_ocns, new_ocns=[])` and aliasing `move_needs_recluster?` got no complaints from rspec.

**Review by jsteverman — CHANGES_REQUESTED** [review]
If my comment was correct about `update_needs_recluster?` and `move_needs_recluster?` being functionally identical then they should be boiled down to `needs_recluster?`

If I'm wrong, we need to do something to clarify the difference or they're going to confuse me for an afternoon in a year or two.

**Pull request merged**